ambry-router/src/main/java/com.github.ambry.router/CryptoWorker.java

```diff
@@ -0,0 +1,112 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Thread responsible for encrypting and decrypting blob content and per-blob keys.
+ * Worker Thread listens to a job queue for new jobs and processes it one by one.
+ */
+public class CryptoWorker implements Runnable, Closeable {
+  private final BlockingQueue<CryptoJob> jobQueue;
+  private final CryptoService cryptoService;
+  private final KeyManagementService kms;
+  private final AtomicBoolean enabled;
+  private volatile boolean endMarkerSeen = false;
+  private final String threadName;
+  private CountDownLatch shutdownLatch;
+
+  private static final Logger logger = LoggerFactory.getLogger(CryptoWorker.class);
+
+  /**
+   * Instantiates {@link CryptoWorker} with the job queue, {@link CryptoService} and {@link KeyManagementService}
+   * @param jobQueue the {@link BlockingQueue} that contains the jobs
+   * @param cryptoService the {@link CryptoService} to use to encrypt or decrypt
+   * @param kms the {@link KeyManagementService} to fetch keys from
+   * @param threadName name of the thread
+   */
+  CryptoWorker(BlockingQueue<CryptoJob> jobQueue, CryptoService cryptoService, KeyManagementService kms,
+      String threadName) {
+    this.jobQueue = jobQueue;
+    this.cryptoService = cryptoService;
+    this.kms = kms;
+    enabled = new AtomicBoolean(true);
+    shutdownLatch = new CountDownLatch(1);
+    this.threadName = threadName;
+  }
+
+  @Override
+  public void run() {
+    try {
+      while (enabled.get()) {
+        try {
+          CryptoJob cryptoJob = jobQueue.take();
+          if (!(cryptoJob instanceof CryptoJob.EndMarkerJob)) {
+            cryptoJob.doOperation(cryptoService, kms);
+          } else {
+            endMarkerSeen = true;
+          }
+        } catch (InterruptedException e) {
+          logger.error("InterruptedException thrown in CryptoWorker");
+        }
+      }
+      if (!endMarkerSeen) {
+        closePendingJobs(new GeneralSecurityException("Shutting down EncryptDecrypt Thread"));
+      }
+    } finally {
+      shutdownLatch.countDown();
+    }
+  }
+
+  /**
+   * Shuts down the thread by closing all pending jobs. Any jobs added after close will be ignored
+   */
+  @Override
+  public void close() {
+    if (enabled.compareAndSet(true, false)) {
+      jobQueue.add(new CryptoJob.EndMarkerJob());
+    }
+    try {
+      shutdownLatch.await(1, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      logger.error("InterruptedException thrown while waiting for CryptoWorker to shutdown");
+    }
+  }
+
+  /**
+   * Closes all pending jobs with the given {@code e}
+   * @param gse the {@link GeneralSecurityException} that needs to be set while invoking callbacks for the pending jobs
+   */
+  private void closePendingJobs(GeneralSecurityException gse) {
+    CryptoJob cryptoJob = jobQueue.poll();
```

Comment:
BlockingQueue actually has a `drainTo` method that might be useful here.
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html#drainTo-java.util.Collection-

Reply:
Good to know this api. I wasn't aware. Unfortunately, we have single queue shared by N no of worker threads. I will try to see if we can incorporate somehow